h/services/nipsa.py

```diff
@@ -15,18 +15,38 @@ class NipsaService(object):
     def __init__(self, session):
         self.session = session
 
-    @property
-    def flagged_userids(self):
+        # Cache of all userids which have been flagged.
+        self._flagged_userids = None
+
+    def fetch_all_flagged_userids(self):
         """
-        A list of all the NIPSA'd userids.
+        Fetch the userids of all shadowbanned / NIPSA'd users.
+
+        The set of userids is cached to speed up subsequent `flagged_userids`
+        and `is_flagged` calls in the same request.
 
         :rtype: set of unicode strings
         """
+        if self._flagged_userids is not None:
+            return self._flagged_userids
+
+        # nb. The `nipsa` field is currently not indexed so this requires a
+        # full table scan.
         query = self.session.query(User).filter_by(nipsa=True)
-        return set([u.userid for u in query])
+        self._flagged_userids = set([u.userid for u in query])
+
+        return self._flagged_userids
 
     def is_flagged(self, userid):
         """Return whether the given userid is flagged as "NIPSA"."""
+
+        # Use the cache if populated.
+        if self._flagged_userids is not None:
```

Comment:
The `is not None` part is needed here because `self._flagged_userids` may have been populated, but may still be an empty set if no users have been shadowbanned.

Reply:
I had a teammate in the past who was adamant about being explicit with the is None's and who knows maybe it ended up saving us from cases like this!